common/src/lib.rs

```diff
@@ -68,6 +68,31 @@ pub struct Program {
     pub nonce: u64,
 }
 
+/// Code with metadata
+///
+/// Storing a code with `submit_code` extrinsic violates code invariant: we no longer
+/// have any guarantees that saved code has an initialized program referencing it (no matter
+/// initialization failed or not). We decided to store `author` and `block_number` as metadata
+/// to the storage for a future opportunity to invalidate some `code` in storage data using either
+/// TTL with `block_number` check or collecting taxes from `author` for storing uninitialized code.
+#[derive(Clone, Debug, Decode, Encode, PartialEq, TypeInfo)]
+pub struct CodeWithMetadata {
+    pub author: H256,
+    pub block_number: u32,
+    pub code: Vec<u8>,
```

Comment:
What is the benefit of storing metadata together with the code? We kinda already have some "metadata" for each piece of code - the reference counter which is in one-to-one relation with the code. Maybe it would make more sense to extend the ref counter (like, replace it with some metadata structure). That way we could access the metadata part independently of the code which might be a bit more efficient considering a piece of code can weigh hundreds of kilobytes so we might want to avoid fetching the whole record from storage every time we need to check, say, TTL expiry or whatnot.

Reply:
Very nice point. Agree.